packages/fx-core/src/plugins/resource/cicd/index.ts

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import {
+  err,
+  UserError,
+  SystemError,
+  v2,
+  TokenProvider,
+  FxError,
+  Result,
+  Inputs,
+  Json,
+  Func,
+  ok,
+  QTreeNode,
+  ProjectSettings,
+  SingleSelectQuestion,
+  Platform,
+} from "@microsoft/teamsfx-api";
+
+import { FxResult, FxCICDPluginResultFactory as ResultFactory } from "./result";
+import { CICDImpl } from "./plugin";
+import { ErrorType, InternalError, NoProjectOpenedError, PluginError } from "./errors";
+import { LifecycleFuncNames } from "./constants";
+import { Service } from "typedi";
+import { ResourcePluginsV2 } from "../../solution/fx-solution/ResourcePluginContainer";
+import { ResourcePlugin, Context, DeepReadonly } from "@microsoft/teamsfx-api/build/v2";
+import {
+  githubOption,
+  azdoOption,
+  jenkinsOption,
+  ciOption,
+  cdOption,
+  provisionOption,
+  publishOption,
+  questionNames,
+} from "./questions";
+import { Logger } from "./logger";
+import { environmentManager } from "../../../core/environment";
+import { telemetryHelper } from "./utils/telemetry-helper";
+
+@Service(ResourcePluginsV2.CICDPlugin)
+export class CICDPluginV2 implements ResourcePlugin {
+  name = "fx-resource-cicd";
+  displayName = "CICD";
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  activate(projectSettings: ProjectSettings): boolean {
+    return true;
+  }
+
+  public cicdImpl: CICDImpl = new CICDImpl();
+
+  public async addCICDWorkflows(
+    context: Context,
+    inputs: Inputs,
+    envInfo: v2.EnvInfoV2
+  ): Promise<Result<any, FxError>> {
+    Logger.setLogger(context.logProvider);
+    return await this.cicdImpl.addCICDWorkflows(context, inputs, envInfo);
+  }
+
+  public async getQuestionsForUserTask(
+    ctx: Context,
+    inputs: Inputs,
+    func: Func,
+    envInfo: DeepReadonly<v2.EnvInfoV2>,
+    tokenProvider: TokenProvider
+  ): Promise<Result<QTreeNode | undefined, FxError>> {
+    const cicdWorkflowQuestions = new QTreeNode({
+      type: "group",
+    });
+
+    const whichProvider = new QTreeNode({
+      name: questionNames.Provider,
+      type: "singleSelect",
+      staticOptions: [githubOption, azdoOption, jenkinsOption],
+      title: "Select a CI/CD Provider",
+      default: githubOption.id,
+    });
+
+    const whichTemplate = new QTreeNode({
+      name: questionNames.Template,
+      type: "multiSelect",
+      staticOptions: [ciOption, cdOption, provisionOption, publishOption],
+      title: "Select template(s)",
+      default: [ciOption.id],
+    });
+
+    if (inputs.platform === Platform.VSCode) {
```

Comment:
Why this only take effect on VSCode? How can CLI and VS use this feature?

Reply:
VS is under active development and CLI's support for VS projects is not ready, so I prefer to not show any VS related stuff in this PR but add all VS stuff in one more PR in the future.

CLI's flow of questions with VS Code is slightly different caused by existing factors:
1. There already exists a question of `Select an environment` in middleware layer https://github.com/OfficeDev/TeamsFx/blob/306d41b05a4b73519374ed836fefb74fe7bf95bf/packages/fx-core/src/core/middleware/envInfoLoader.ts#L257
2. CICD Plugin returns all three questions including env question to the upper layer for VS Code, so the three questions are in one batch (can return to previous question). That is, the flow didn't reuse the env question in middleware to avoid the separation of env question with the other two ones.
3. CICD Plugin returns two questions (provider, template) to the upper layer for CLI, and leverage the env question from middleware since there's no `go back` experience in CLI.